File: backend/src/services/assessmentPipeline.ts
import type { AssessmentResult, ProjectInput, GeoLocation } from '../types';
import { geocodeAddress, lookupGridOperator } from './geocoding';
import { calcTechnical } from './technicalCalc';
import { retrieveRelevantChunks } from './ragRetrieval';
import { callAssessmentLlm } from './llmService';

export async function runAssessmentPipeline(
  projectId: string,
  input: ProjectInput
): Promise<AssessmentResult> {
  // Step 1: Geocode address
  let geo: GeoLocation | null = null;
  try {
    geo = await geocodeAddress(input.address);
  } catch (e) {
    console.warn('Geocoding failed:', e);
  }

  // Step 2: Grid operator lookup
  let gridOperator = null;
  if (geo) {
    try {
      gridOperator = await lookupGridOperator(geo.lat, geo.lng);
    } catch (e) {
      console.warn('Grid operator lookup failed:', e);
    }
  }

  // Enrich input with resolved operator id
  if (gridOperator) {
    input = { ...input, gridOperatorId: gridOperator.id };
  }

  // Step 3: Technical pre-calculation
  const technicalSummary = calcTechnical(input);

  // Step 4 & 5: RAG retrieval
  const chunks = await retrieveRelevantChunks(input);

  // Step 6: LLM call
  let llmResult: Partial<AssessmentResult> & { rawLlmTrace: string };
  try {
    llmResult = await callAssessmentLlm(input, technicalSummary, gridOperator, chunks);
  } catch (e) {
    console.error('LLM call failed, using fallback:', e);
    // Fallback: return a minimal valid result if LLM is unavailable
    llmResult = buildFallbackResult();
  }

  // Step 7: Assemble final result
  const result: AssessmentResult = {
    projectId,
    generatedAt: new Date().toISOString(),
    gridOperator,
    technicalSummary,
    regulatoryClaims: llmResult.regulatoryClaims ?? [],
    subsidies: llmResult.subsidies ?? [],
    openPoints: llmResult.openPoints ?? [],
    nextSteps: llmResult.nextSteps ?? [],
    installerQuestions: llmResult.installerQuestions ?? [],
    trafficLight: llmResult.trafficLight ?? 'amber',
    rawLlmTrace: llmResult.rawLlmTrace,
  };

  return result;
}

function buildFallbackResult(): Partial<AssessmentResult> & { rawLlmTrace: string } {
  return {
    regulatoryClaims: [
      {
        text: 'MaStR-Registrierung (Pflicht)',
        detail: 'Binnen 1 Monat nach Inbetriebnahme · marktstammdatenregister.de',
        sourceRef: '§§3 Nr.1, 5 MaStRV · seit 01.07.2017',
        sourceIds: ['mastr-rv'],
        status: 'valid',
      },
      {
        text: 'Netzanmeldung vor Installation (Pflicht)',
        detail: 'Mindestens 4 Wochen vor Inbetriebnahme beim Netzbetreiber anzeigen',
        sourceRef: '§13 NAV · gültig',
        sourceIds: ['nav-13'],
        status: 'valid',
      },
    ],
    subsidies: [
      {
        name: 'KfW 270 — Kredit',
        shortName: 'KfW 270',
        status: 'valid',
        amount: 'bis 150.000 EUR',
        description: '5–30 Jahre · über Hausbank · vor Beauftragung beantragen!',
      },
    ],
    openPoints: [
      'LLM-Analyse nicht verfügbar — bitte API-Key prüfen',
      'Netzanschlusskapazität beim Netzbetreiber anfragen',
    ],
    nextSteps: [
      { text: 'Netzanschlussanfrage beim Netzbetreiber stellen', priority: 'high' },
      { text: 'KfW 270 über Hausbank beantragen (vor Beauftragung!)', priority: 'high' },
    ],
    installerQuestions: [
      'Welche technischen Anforderungen gelten für diesen Netzanschluss?',
      'Wie wird die 70%-Wirkleistungsbegrenzung nach VDE-AR-N 4105 umgesetzt?',
    ],
    trafficLight: 'amber',
    rawLlmTrace: 'FALLBACK — LLM unavailable',
  };
}

File: backend/app/services/assessment_pipeline.py
"""Assessment pipeline — port of services/assessmentPipeline.ts."""
from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import Any, Dict
from uuid import uuid4

from app.models import AssessmentResult, Claim, GeoLocation, GridOperator, Priority, ProjectInput, RegulatoryStatus, Step, Subsidy, TechnicalSummary, TrafficLight
from app.services.geocoding import geocode_address, lookup_grid_operator
from app.services.technical_calc import calc_technical
from app.services.pvgis import get_irradiance
from app.services.rag_retrieval import retrieve_relevant_chunks
from app.services.llm_service import _build_fallback_result, call_assessment_llm

log = logging.getLogger(__name__)


def _coerce_claims(raw: list[dict]) -> list[Claim]:
    result = []
    for c in raw:
        try:
            result.append(Claim(**c))
        except Exception:
            pass
    return result


def _coerce_subsidies(raw: list[dict]) -> list[Subsidy]:
    result = []
    for s in raw:
        try:
            result.append(Subsidy(**s))
        except Exception:
            pass
    return result


def _coerce_steps(raw: list[dict]) -> list[Step]:
    result = []
    for s in raw:
        try:
            result.append(Step(**s))
        except Exception:
            pass
    return result


async def run_assessment_pipeline(project_id: str, inp: ProjectInput) -> AssessmentResult:
    # Step 1: Geocode address
    geo: GeoLocation | None = None
    try:
        geo = await geocode_address(inp.address)
    except Exception as e:
        log.warning("Geocoding failed: %s", e)

    # Step 2: Grid operator lookup
    grid_operator: GridOperator | None = None
    if geo:
        try:
            grid_operator = await lookup_grid_operator(geo.lat, geo.lng)
        except Exception as e:
            log.warning("Grid operator lookup failed: %s", e)

    # Enrich input with resolved operator id
    if grid_operator:
        inp = inp.model_copy(update={"gridOperatorId": grid_operator.id})

    # Step 3: Technical pre-calculation (with PVGIS irradiance)
    irradiance_val = 950.0
    if geo:
        try:
            irr_data = await get_irradiance(
                geo.lat, geo.lng, inp.roofAreaM2, inp.roofOrientation.value, inp.roofPitchDeg
            )
            irradiance_val = irr_data.get("irradianceKwhPerKwp", 950.0)
        except Exception as e:
            log.warning("Irradiance lookup failed: %s", e)

    technical_summary = calc_technical(inp, irradiance_val)

    # Step 4 & 5: RAG retrieval
    chunks = await retrieve_relevant_chunks(inp)

    # Step 6: LLM call
    try:
        llm_result = await call_assessment_llm(inp, technical_summary, grid_operator, chunks)
    except Exception as e:
        log.error("LLM call failed, using fallback: %s", e)
        llm_result = _build_fallback_result()

    # Step 7: Assemble final result
    return AssessmentResult(
        projectId=project_id,
        generatedAt=datetime.now(timezone.utc).isoformat(),
        gridOperator=grid_operator,
        technicalSummary=technical_summary,
        regulatoryClaims=_coerce_claims(llm_result.get("regulatoryClaims", [])),
        subsidies=_coerce_subsidies(llm_result.get("subsidies", [])),
        openPoints=llm_result.get("openPoints", []),
        nextSteps=_coerce_steps(llm_result.get("nextSteps", [])),
        installerQuestions=llm_result.get("installerQuestions", []),
        trafficLight=TrafficLight(llm_result.get("trafficLight", "amber")),
        rawLlmTrace=llm_result.get("rawLlmTrace"),
    )
